src/Razor/src/Microsoft.CodeAnalysis.Remote.Razor/RemoteMefComposition.cs

```diff
@@ -52,17 +61,99 @@ private static async Task<CompositionConfiguration> CreateConfigurationAsync(Can
     ///  are reused for subsequent calls to this method.
     /// </summary>
-    public static async Task<ExportProvider> CreateExportProviderAsync(CancellationToken cancellationToken)
+    public static async Task<ExportProvider> CreateExportProviderAsync(string? cacheDirectory, CancellationToken cancellationToken)
     {
+        var cache = new CachedComposition();
+        var compositionCacheFile = GetCompositionCacheFile(cacheDirectory);
+        if (await TryLoadCachedExportProviderAsync(cache, compositionCacheFile, cancellationToken).ConfigureAwait(false) is { } cachedProvider)
+        {
+            return cachedProvider;
+        }
+
         var configuration = await s_lazyConfiguration.GetValueAsync(cancellationToken).ConfigureAwait(false);
         cancellationToken.ThrowIfCancellationRequested();
 
         var runtimeComposition = RuntimeComposition.CreateRuntimeComposition(configuration);
         var exportProviderFactory = runtimeComposition.CreateExportProviderFactory();
 
+        // We don't need to block on saving the cache, because if it fails or is corrupt, we'll just try again next time, but
+        // we capture the task just so that tests can verify things.
+        s_saveCacheFileTask = TrySaveCachedExportProviderAsync(cache, compositionCacheFile, runtimeComposition, cancellationToken);
+
         return exportProviderFactory.CreateExportProvider();
     }
 
+    private static async Task<ExportProvider?> TryLoadCachedExportProviderAsync(CachedComposition cache, string? compositionCacheFile, CancellationToken cancellationToken)
+    {
+        if (compositionCacheFile is null)
+        {
+            return null;
+        }
+
+        try
+        {
+            if (File.Exists(compositionCacheFile))
+            {
+                var resolver = new Resolver(SimpleAssemblyLoader.Instance);
+                using var cacheStream = new FileStream(compositionCacheFile, FileMode.Open, FileAccess.Read, FileShare.Read, bufferSize: 4096, useAsync: true);
+                var cachedFactory = await cache.LoadExportProviderFactoryAsync(cacheStream, resolver, cancellationToken).ConfigureAwait(false);
+                return cachedFactory.CreateExportProvider();
+            }
+        }
+        catch (Exception)
+        {
+            // We ignore all errors when loading the cache, because if the cache is corrupt we will just create a new export provider.
+        }
+
+        return null;
+    }
+
+    private static async Task TrySaveCachedExportProviderAsync(CachedComposition cache, string? compositionCacheFile, RuntimeComposition runtimeComposition, CancellationToken cancellationToken)
+    {
+        if (compositionCacheFile is null)
+        {
+            return;
+        }
+
+        try
+        {
+            var cacheDirectory = Path.GetDirectoryName(compositionCacheFile).AssumeNotNull();
+            var directoryInfo = Directory.CreateDirectory(cacheDirectory);
```

Comment:
>var directoryInfo = Directory.CreateDirectory(cacheDirectory); [](http://example.com/codeflow?start=12&length=62)

No concern about cleanup of old caches?

Reply:
🤦‍♂️ Thanks. This was absolutely on my todo list for this PR, but I guess I forgot when I got distracted by the worry about perf.

Definitely need to clean up in VS. VS Code should be fine because each extension version is a separate folder, but I don't think there is any harm in doing it regardless.